src/v/serde/parquet/writer.cc

```diff
@@ -0,0 +1,146 @@
+#include "serde/parquet/column_writer.h"
+#include "serde/parquet/metadata.h"
+#include "serde/parquet/shredder.h"
+
+namespace serde::parquet {
+
+class writer::impl {
+public:
+    impl(options opts, ss::output_stream<char> output)
+      : _opts(std::move(opts))
+      , _output(std::move(output)) {}
+
+    ss::future<> init() {
+        index_schema(_opts.schema);
+        _opts.schema.for_each([this](const schema_element& element) {
+            if (!element.is_leaf()) {
+                return;
+            }
+            _column_writers.emplace(element.position, element);
+            _schema_leaves.emplace(element.position, &element);
+        });
+        // write the leading magic bytes
+        co_await write_iobuf(iobuf::from("PAR1"));
+    }
+
+    ss::future<> write_row(group_value row) {
+        co_await shred_record(
+          _opts.schema, std::move(row), [this](shredded_value sv) {
+              return write_value(std::move(sv));
+          });
+        // TODO: periodically flush the row_group if we're using enough memory
+        // or at least flush the pages if we don't want to create too many row
+        // groups.
+    }
+
+    ss::future<> close() {
+        chunked_vector<row_group> row_groups;
+        row_groups.push_back(co_await flush_row_group());
+        int64_t num_rows = 0;
+        for (const auto& rg : row_groups) {
+            num_rows += rg.num_rows;
+        }
+        auto encoded_footer = encode(file_metadata{
+          .version = 2,
+          .schema = flatten(_opts.schema),
+          .num_rows = num_rows,
+          .row_groups = std::move(row_groups),
+          .key_value_metadata = std::move(_opts.metadata),
+          .created_by = fmt::format(
+            "Redpanda version {} (build {})", _opts.version, _opts.build),
+        });
+        size_t footer_size = encoded_footer.size_bytes();
+        co_await write_iobuf(std::move(encoded_footer));
+        co_await write_iobuf(encode_footer_size(footer_size));
+        co_await write_iobuf(iobuf::from("PAR1"));
+        co_await _output.close();
+    }
+
+private:
+    iobuf encode_footer_size(size_t size) {
+        iobuf b;
+        auto le_size = ss::cpu_to_le(static_cast<uint32_t>(size));
+        // NOLINTNEXTLINE(*reinterpret-cast*)
+        b.append(reinterpret_cast<const uint8_t*>(&le_size), sizeof(le_size));
+        return b;
+    }
+
+    ss::future<> write_value(shredded_value sv) {
+        auto& col = _column_writers.at(sv.schema_element_position);
+        col.add(std::move(sv.val), sv.rep_level, sv.def_level);
+        return ss::now();
+    }
+
+    ss::future<row_group> flush_row_group() {
+        row_group rg{};
+        rg.file_offset = static_cast<int64_t>(_offset);
+        for (auto& [pos, col] : _column_writers) {
+            const schema_element* leaf = _schema_leaves.at(pos);
+            auto page = col.flush_page();
+            const auto& data_header = std::get<data_page_header>(
```

Comment:
In the current form we have 1 row group and one page per column within the group, right?

Reply:
Correct.